IndexSearch/pinot-label.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH PINOT-LABEL "1" "May 2026" "pinot 1.24" "User Commands"
.SH NAME
pinot-label \- Label files from the command-line
.SH SYNOPSIS
.B pinot-label
[\fI\,OPTIONS\/\fR] [\fI\,FILES\/\fR]
.SH DESCRIPTION
pinot\-label \- Label files from the command\-line
.SH OPTIONS
.TP
\fB\-g\fR, \fB\-\-get\fR
get the labels list for the given file
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-l\fR, \fB\-\-list\fR
list known labels
.TP
\fB\-r\fR, \fB\-\-reload\fR
get the daemon to reload the configuration
.TP
\fB\-s\fR, \fB\-\-set\fR
set labels on the given file
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.SH EXAMPLES
pinot\-label \-\-get /home/fabrice/Documents/Bozo.txt
.PP
pinot\-label \-\-list
.PP
pinot\-label \-\-set "[Clowns][Fun][My Hero]" /home/fabrice/Documents/Bozo.txt
.SH "REPORTING BUGS"
Report bugs to fabrice.colin@gmail.com
.PP
.br
This is free software.  You may redistribute copies of it under the terms of
the GNU General Public License <http://www.gnu.org/licenses/old\-licenses/gpl\-2.0.html>.
.br
There is NO WARRANTY, to the extent permitted by law.


NEWS
2026/05/02 version_1_2_4
General :
 - switched to C++17
Collect :
 - Neon-based downloader was lagging behind

2025/11/09 version_1_2_3
Collect :
 - curl 8.16 type fix
 - the HTML filter supports anchors titles
IndexSearch :
 - updated the Wikipedia plugin
UI :
 - default 64x64 icon

2024/10/27 version_1_2_2
General :
 - const correctness and Debian unstable build fixes
 - support for exiv2 0.28
 - require libarchive >= 3.0.0
 - require libxml++ >= 3.2.0
 - several warning fixes

2022/02/22 version_1_2_1
General :
 - removed code paths around diacritics sensitivity
 - removed code paths pertaining to other query languages
 - auto-generate the ChangeLog file
IndexSearch :
 - removed historical checks on old Xapian releases
 - replaced homegrown abstract solution with Xapian's
UI :
 - fixed query history

2021/10/16 version_1_2_0
General :
 - dropped xdgmime in favour of gio
 - require giomm >= 2.6
 - run rst2txt on RST files, if available
IndexSearch :
 - removed obsolete search plugins, dead code
 - added a plugin for Arxiv
 - better results with the dir query filter
 - fixed size range queries
 - synced with how Xapian Omega indexes file extensions
 - added sort by file size (descending)
 - pinot-index shows document IDs to facilitate troubleshooting
 with xapian-delve
 - require Xapian >= 1.4.10
 - reindexing is recommended.
Daemon :
 - fixed a bug that would cause the loss of documents for one directory
 when crawling ends for the next one
 - overhauled the D-Bus implementation, extended the interface
 - dropped dependency on dbus-glib
 - reduced flushing
 - a IndexFlushEpoch property indicates the last time index changed
 were flushed to disk
 - rely on org.freedesktop.UPower's OnBattery on systems with a battery
 - implemented the org.gnome.Shell.SearchProvider2 interface. The content
 of documents in directories indexed by the daemon can be searched through
 the Gnome 3 Shell.
UI :
 - removed the gtk2 UI, refreshed the gtk3 UI
 - the gtk3 UI is loaded from a Glade XML file
 - reduced direct reading of the daemon's index
 - reopen the daemon's index as per the IndexFlushEpoch property
 - fixed file imports and error reporting
 - stored queries' results may be sorted by file size
 - rationalized activation and behaviour of the Edit menuitems
 - pinot may be started on any new query with -q/--query TERMS
 - require gtkmm >= 3.24

2020/01/09 version_1_1_0
General:
 - updated gmime, textcat, exiv, glib, OpenSSL and Xapian dependencies

2015/06/11 version_1_0_9
Tokenize :
 - new JSON filter
 - streamlined the mbox filter a bit
 - pinot-index looks for filters in the same locations as the UI
IndexSearch :
 - sort documents by date, in ascending or descending order

2014/07/18 version_1_0_8
General :
 - don't install the xdgmime files
 - use libnotify if --enable-libnotify=yes is specified
 - fixed check on library symbols that could cause a crash, added
 support for LLVM-built filters
 - general clean up
SQL :
 - refactored the DB interface

2014/05/22 version_1_0_7
General:
 - compilation fixes
 - prefer off_t for file sizes, offsets, buffer lengths
Monitor :
 - try and reapply watches on directories that have just been removed
Tokenize :
 - refactored encoding conversion
IndexSearch :
 - fixed build with libexttextcat

2013/05/26 version_1_0_6
IndexSearch:
 - support boost 1.50's Spirit
SQL :
 - better handling of potential errors while stepping through results
Collect :
 - minor fixes to curl backend

2013/03/03 version_1_0_5
IndexSearch:
 - fixes to abstract generation, CJKV tokenization

2013/02/11 version_1_0_4
IndexSearch:
 - fix stripping of diacritics
 - stem subject terms

2013/01/14 version_1_0_3
IndexSearch :
 - fix a Unicode handling issue introduced in 1.01
UI :
 - updated French translation by Eliovir

2012/11/04 version_1_0_2
General :
 - turn memory pooling off by default to avoid issues with newer boost
UI :
 - new Czech translation by Zbyněk Schwarz
 - updated Japanese translation by Takafumi Arakaki
 - updated Brazilian Portuguese translation by Adriano Steffler

2012/08/27 version_1_0_1
General :
 - run rst2html on RST files, if available, and if RST files are detected as such.
 See http://code.google.com/p/pinot-search/issues/detail?id=12
Tokenize :
 - better mbox parts extraction
IndexSearch :
 - dropped unac in favour of own code, resulting in faster indexing.
 - pinot-index --override MIMETYPE:EXTENSION overrides MIME type detection
 based on files extensions

2012/06/16 version_1_0_0
General :
 - install headers and libraries
 - prefer default programs that don't support URIs to view local files
Deskbar :
 - dropped support for Deskbar since it's now dead
Tokenize :
 - better mbox parsing. The internal part numbering scheme has changed;
 reindexing email is recommended.
IndexSearch :
 - support for LibreOffice's libexttextcat v3.2
 - fixed the Google plugin
 - renamed the Freshmeat plugin to Freecode
UI :
 - updated Simplified Chinese translation by happymeng
 - updated German translation by Gena Haltmair
 - initial GTK+ 3 port. Enable with "./configure ... --enable-gtkmm3=yes"

2011/11/07 version_0_9_8
Tokenize :
 - new exiv2-based filter
 - new chmlib-based filter
IndexSearch :
 - support for LibreOffice's libexttextcat v3.1, and possibly v3.1.1
 - dropped plugins for Yahoo! REST API, Yahoo! BOSS, Google Code Search
  and RollYO
UI :
 - query results and view history are expired after 6 months
 - updated Dutch translation by Martijn Verstrate and Tico
 - updated German translation by Fitoschido
 - updated Russian translation by Nikolay Kachanov
 - updated Spanish translation by pkramerruiz

2011/01/09 version_0_9_7
General :
 - replaced custom memory pool class with Boost's
 - don't try and map more than 2Gb and use shared mappings
Deskbar :
 - install the module where DeskBar > 2.28 expects it to be
IndexSearch :
 - remove dots at the end of terms that don't look like acronyms
 - index components of acronyms and dot-separated terms on their own
Daemon :

IndexSearch/pinot-label.cpp
/*
 *  Copyright 2007-2021 Fabrice Colin
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
 
#include <getopt.h>
#include <stdlib.h>
#include <iostream>
#include <string>
#include <set>
#include <giomm/init.h>
#include <glibmm/init.h>

#include "config.h"
#include "StringManip.h"
#include "MIMEScanner.h"
#include "Url.h"
#include "DBusIndex.h"

using namespace std;

static struct option g_longOptions[] = {
	{"get", 0, 0, 'g'},
	{"help", 0, 0, 'h'},
	{"list", 0, 0, 'l'},
	{"reload", 0, 0, 'r'},
	{"set", 1, 0, 's'},
	{"version", 0, 0, 'v'},
	{0, 0, 0, 0}
};


static void printLabels(const set<string> &labels, const string &fileName)
{
	if (fileName.empty() == false)
	{
		clog << fileName << endl;
	}
	clog << "Labels: ";

	for (set<string>::const_iterator labelIter = labels.begin();
		labelIter != labels.end(); ++labelIter)
	{
		if (labelIter->substr(0, 2) == "X-")
		{
			continue;
		}
		clog << "[" << Url::escapeUrl(*labelIter) << "]";
	}
	clog << endl;
}

static void printHelp(void)
{
	// Help
	clog << "pinot-label - Label files from the command-line\n\n"
		<< "Usage: pinot-label [OPTIONS] [FILES]\n\n"
		<< "Options:\n"
		<< "  -g, --get                 get the labels list for the given file\n"
		<< "  -h, --help                display this help and exit\n"
		<< "  -l, --list                list known labels\n"
		<< "  -r, --reload              get the daemon to reload the configuration\n"
		<< "  -s, --set                 set labels on the given file\n"
		<< "  -v, --version             output version information and exit\n\n";
	clog << "Examples:\n"
		<< "pinot-label --get /home/fabrice/Documents/Bozo.txt\n\n"
		<< "pinot-label --list\n\n"
		<< "pinot-label --set \"[Clowns][Fun][My Hero]\" /home/fabrice/Documents/Bozo.txt\n\n"
		<< "Report bugs to " << PACKAGE_BUGREPORT << endl;
}

int main(int argc, char **argv)
{
	set<string> labels;
	string labelsString;
	int longOptionIndex = 0;
	unsigned int docId = 0;
	int minArgNum = 1;
	bool getLabels = false, getDocumentLabels = false, reloadIndex = false, setDocumentLabels = false, success = false;

	// Look at the options
	int optionChar = getopt_long(argc, argv, "ghlrs:v", g_longOptions, &longOptionIndex);
	while (optionChar != -1)
	{
		set<string> engines;

		switch (optionChar)
		{
			case 'g':
				getDocumentLabels = true;
				break;
			case 'h':
				printHelp();
				return EXIT_SUCCESS;
			case 'l':
				minArgNum = 0;
				getLabels = true;
				break;
			case 'r':
				minArgNum = 0;
				reloadIndex = true;
				break;
			case 's':
				setDocumentLabels = true;
				if (optarg != NULL)
				{
					labelsString = optarg;
				}
				break;
			case 'v':
				clog << "pinot-label - " << PACKAGE_STRING << "\n\n"
					<< "This is free software.  You may redistribute copies of it under the terms of\n"
					<< "the GNU General Public License <http://www.gnu.org/licenses/old-licenses/gpl-2.0.html>.\n"
					<< "There is NO WARRANTY, to the extent permitted by law." << endl;
				return EXIT_SUCCESS;
			default:
				return EXIT_FAILURE;
		}

		// Next option
		optionChar = getopt_long(argc, argv, "ghls:v", g_longOptions, &longOptionIndex);
	}

	if (argc == 1)
	{
		printHelp();
		return EXIT_SUCCESS;
	}

	if ((argc < 2) ||
		(argc - optind < minArgNum))
	{
		clog << "Not enough parameters" << endl;
		return EXIT_FAILURE;
	}

	if ((setDocumentLabels == true) &&
		(labelsString.empty() == true))
	{
		clog << "Incorrect parameters" << endl;
		return EXIT_FAILURE;
	}

	Glib::init();
	Gio::init();
	// Initialize GType
#if !GLIB_CHECK_VERSION(2,35,0)
	g_type_init();
#endif

	MIMEScanner::initialize("", "");

	// We need a pure DBusIndex object
	DBusIndex index(NULL);

	if (getLabels == true)
	{
		if (index.getLabels(labels) == true)
		{
			printLabels(labels, "");

			success = true;
		}
	}

	while (optind < argc)
	{
		string fileParam(argv[optind]);
		Url thisUrl(fileParam, "");

		// Rewrite it as a local URL
		string urlParam(thisUrl.getProtocol());
		urlParam += "://";
		urlParam += thisUrl.getLocation();
		if (thisUrl.getFile().empty() == false)
		{
			urlParam += "/";
			urlParam += thisUrl.getFile();
		}
#ifdef DEBUG
		clog << "URL rewritten to " << urlParam << endl;
#endif

		if ((getDocumentLabels == true) ||
			(setDocumentLabels == true))
		{
			docId = index.hasDocument(urlParam);
			if (docId == 0)
			{
				clog << fileParam << " is not indexed" << endl;
				success = false;

				// Next
				++optind;
				continue;
			}
		}

		if (getDocumentLabels == true)
		{
			labels.clear();

			if (index.getDocumentLabels(docId, labels) == true)
			{
				printLabels(labels, fileParam);

				success = true;
			}
		}

		if (setDocumentLabels == true)
		{
			string::size_type endPos = 0;
			string label(StringManip::extractField(labelsString, "[", "]", endPos));

			labels.clear();

			// Parse labels
			while (label.empty() == false)
			{
				labels.insert(Url::unescapeUrl(label));

				if (endPos == string::npos)
				{
					break;
				}
				label = StringManip::extractField(labelsString, "[", "]", endPos);
			}

#ifdef DEBUG
			printLabels(labels, fileParam);
#endif
			success = index.setDocumentLabels(docId, labels);
		}

		// Next
		++optind;
	}

	if (reloadIndex == true)
	{
		index.reload();
	}

	MIMEScanner::shutdown();

	// Did whatever operation we carried out succeed ?
	if (success == true)
	{
		return EXIT_SUCCESS;
	}

	return EXIT_FAILURE;
}
